Handle unbacked symints in buffer reuse calculation

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109609
* __->__ #109603

This is rewritten from https://github.com/pytorch/pytorch/pull/106655 to land faster, with peterbell10's comments.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ezyang committed:
Handle unbacked symints in buffer reuse calculation

This is extracted from https://github.com/pytorch/pytorch/pull/106655
to land faster.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang commented:
@yf225 if you'd prefer to retain authorship feel free to resubmit this PR as you. I just want this landed ASAP

peterbell10 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -35,13 +36,24 @@
         node.get_dtype(),
         V.graph.sizevars.simplify(sympy_product(size)),
-        # Detect gaps in tensor storage caused by strides
-        V.graph.sizevars.size_hint(last_element),
+        size_hint_of_last_element,
```
I might be missing something but I think these two conditions should be replaced by

```python
        V.graph.sizevars.simplify(node.layout.storage_size()),
```

1. It's an equivalent calculation to `last_element` except it's a symbolic quantity, so there is no issue with unbacked symint.
2. if there are gaps in the layout, the numel won't match so we wouldn't end up reusing buffers if we checked both match

ezyang replied:
I agree that the `storage_size` quantity is equivalent as it calls 

```
def compute_required_storage_length(
    shape: ShapeType, strides: StrideType, storage_offset: int
) -> int:
    """Computes the minimum storage size to hold the given tensor geometry.

    Example
    =======

    This is the size of a newly allocated tensor's storage, in units of elements

    >>> t = torch.empty((10, 20))
    >>> compute_required_storage_length(t.shape, t.stride(), t.storage_offset())
    200

    >>> # xdoctest: +SKIP(failing)
    >>> t2 = torch.empty_strided((1, 2, 3), (5, 7, 11))
    >>> size = compute_required_storage_length(t2.shape, t2.stride(), t2.storage_offset())
    >>> size == t.storage().size()
    True

    A valid tensor may have a larger storage size, but never smaller

    >>> slice = torch.empty(100)[20:40]
    >>> slice.storage().size()
    100

    >>> compute_required_storage_length(slice.shape, slice.stride(), slice.storage_offset())
    40

    """
    # Short-circuits if the shape has no elements
    if reduce(operator.mul, shape, 1) == 0:
        return 0
    
    max_offset = sum((x - 1) * y for x, y in zip(shape, strides))
    # +1 to account for the first element which offsets are taken from
    return 1 + storage_offset + max_offset
```

However, I think we still are obligated to `sympy_str` the quantity when it is symbolic, because otherwise we'll try to hash on sympy expressions which may or may not work in the way we expect.

peterbell10 replied:
> However, I think we still are obligated to `sympy_str` the quantity when it is symbolic, because otherwise we'll try to hash on sympy expressions which may or may not work in the way we expect.

The line above uses the symbolic form of numel and that seems to work fine. I suppose the stringified version should be equivalent though so I guess it's okay.

ezyang committed:
Update on "Handle unbacked symints in buffer reuse calculation"


This is extracted from https://github.com/pytorch/pytorch/pull/106655
to land faster.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang committed:
Update on "Handle unbacked symints in buffer reuse calculation"


This is extracted from https://github.com/pytorch/pytorch/pull/106655
to land faster.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "spurious failures cancelled it"
